### ISSUE TITLE
Optional certificate verification requires CA chain

### ISSUE THREAD

**Issue by jethrogb** [ARMmbed/mbedtls#864]
If I run a server with `MBEDTLS_SSL_VERIFY_OPTIONAL` but don't set `ca_chain` (to send the CertificateRequest message with an empty DN list, see RFC 5246), and the client sends a certificate, the code in `mbedtls_ssl_parse_certificate` will always fail with `MBEDTLS_ERR_SSL_CA_CHAIN_REQUIRED`. Certificate verification should never fail when using `MBEDTLS_SSL_VERIFY_OPTIONAL`.

**Comment by jethrogb** [comment]
Suggested change:

```patch
diff --git a/vendor/library/ssl_tls.c b/vendor/library/ssl_tls.c
index 806e0ca..8b43c57 100644
--- a/vendor/library/ssl_tls.c
+++ b/vendor/library/ssl_tls.c
@@ -4405,7 +4405,7 @@ int mbedtls_ssl_parse_certificate( mbedtls_ssl_context *ssl )
             ca_crl   = ssl->conf->ca_crl;
         }
 
-        if( ca_chain == NULL )
+        if( ca_chain == NULL && authmode == MBEDTLS_SSL_VERIFY_REQUIRED )
         {
             MBEDTLS_SSL_DEBUG_MSG( 1, ( "got no CA chain" ) );
             return( MBEDTLS_ERR_SSL_CA_CHAIN_REQUIRED );
```

**Comment by sbutcher-arm** [comment]
Thanks for raising the issue!

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-1341

**Comment by RonEld** [comment]
fixed on versions mbedtls-2.5.1 , mbedtls-2.1.8 and mbedtls-1.3.20
